rust_src/src/multibyte.rs

```diff
@@ -88,13 +92,85 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions. This is a convenience function for
+    /// `self.display_width(None)`
     pub fn width(self) -> usize {
-        unsafe { lisp_string_width(self.into(), -1, ptr::null_mut(), ptr::null_mut()) as usize }
+        let (width, _) = self.display_width(None);
+        width
+    }
+
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions.
+    ///
+    /// With `precision` argument, return the width of longest substring that
+    /// doesn't exceed `precision`, and the number of characters and bytes it
+    /// contains in the returned tuple.
+    pub fn display_width(self, precision: Option<usize>) -> (usize, Option<(usize, usize)>) {
+        // Manually determine if string is unibyte.
+        let len = self.len_chars() as usize;
+        let multibyte = self.len_chars() < self.len_bytes();
+        let distab = unsafe { buffer_display_table() };
+        // Sum width
+        let mut width = 0;
+        // Character index
+        let mut i = 0;
+        // Byte index
+        let mut b = 0;
+
+        while i < len {
```

Comment:
Could this be done with more sophisticated control flow? Maybe a for loop with a range or something? That way you could remove the i-mutation statement at the end of the loop. A really slick solution would also get rid of the b-mutation statement. (I don't know how, but that's my feeling.)

Reply:
That's going to be difficult considering that a single iteration can jump not only several bytes, but also several code points. Anything involving iterators is going to be significantly more clumsy than just mutating an index since it'll require a "skip counter" or something similar, putting us back to square one but uglier.